packages/flutter_tools/lib/src/compile.dart

```diff
@@ -1064,6 +1060,38 @@ class DefaultResidentCompiler implements ResidentCompiler {
     server.kill();
     return server.exitCode;
   }
+
+  Future<void> _writelnToServerStdin(String line, {
+    bool printTrace = false,
+  }) async {
+    await _writelnToServerStdinAll(<String>[line], printTrace: printTrace);
+  }
+
+  // TODO(andrewkolos): Concurrent calls to ProcessUtils.writelnToStdinUnsafe
+  //  against the same stdin will result in an exception. To guard against this,
+  //  we need to force calls to run serially. Ideally, this wouldn't be
+  //  necessary since we shouldn't have multiple concurrent writes to the
+  //  compiler process. However, we do (https://github.com/flutter/flutter/issues/152577).
+  final Pool _serverStdinWritePool = Pool(1);
+  Future<void> _writelnToServerStdinAll(List<String> lines, {
+    bool printTrace = false,
+  }) async {
+    final Process? server = _server;
+    if (server == null) {
+      return;
+    }
+    final PoolResource request = await _serverStdinWritePool.request();
+    try {
+      for (final String line in lines) {
+        await ProcessUtils.writelnToStdinUnsafe(stdin: server.stdin, line: line);
```

Comment:
i worry that we're adding a lot of flush calls and asynchronous suspensions here. Would it work if we do:

```
  await ProcessUtils.writelnToStdinUnsafe(stdin: server.stdin, line: lines.join('\n'));
```

Effectively batching the writelns before a single flush call?

Reply:
Hmm, a hot reload test timed out:

```
24:00 +58 -1: test/integration.shard/hot_reload_test.dart: hot restart works without error [E]
  TimeoutException after 0:15:00.000000: Test timed out after 15 minutes.
  dart:isolate  _RawReceivePort._handleMessage
```

Did the batching break the compiler?

Reply:
This may be a manifestation of https://github.com/flutter/flutter/issues/152220 (or something with the same root cause), but now that you point it out...let me see if there's anything in the logs that can support this.

Reply:
I think this is different. I find this from the[ logs](https://ci.chromium.org/ui/p/flutter/builders/try/Mac%20tool_integration_tests_2_5/324/overview) interesting:

```
<=stdout=  [{"event":"app.stop","params":{"appId":"0b7edf75-f1fe-442c-ad22-155775e36df2"}}]
  Exception: Received app.stop event while waiting for response to request 1
```

for reference:

```
Performing hot restart...

=stdin=>   [{"id":1,"method":"app.restart","params":{"appId":"0b7edf75-f1fe-442c-ad22-155775e36df2","fullRestart":true,"pause":false,"debounce":false,"debounceDurationOverrideMs":null}}]
```

So the tool never gets a response from the VM service for the `app.restart` call? Unfortunately it looks like the last message from the VM service is truncated in the logs.